# Data validation between the data set in BigQuery and GraphDB

The objectives are to:
1. Ensure complete upload of data to BigQuery and GraphDB.
2. Check for data qualtiy issues.
3. Gain insights on the data.
4. Engineer features.

In [92]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [93]:
# Mount to Google Drive to save results
from google.colab import drive
drive.mount('/content/drive')
%cd /content/drive/MyDrive/MSc/2020-21/Research\ Project/Colab/
%ls

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
/content/drive/MyDrive/MSc/2020-21/Research Project/Colab


In [94]:
# Connect to GCP Bucket
from google.colab import auth
auth.authenticate_user()

In [95]:
# Set GCP project ID and region to Europe West 2 - London
PROJECT = 'detect-fake-news-313201'
!gcloud config set project $PROJECT
REGION = 'europe-west2'
CLUSTER = '{}-cluster'.format(PROJECT)
!gcloud config set compute/region $REGION
!gcloud config set dataproc/region $REGION

!gcloud config list # show some information

Updated property [core/project].
Updated property [compute/region].
Updated property [dataproc/region].
[component_manager]
disable_update_check = True
[compute]
gce_metadata_read_timeout_sec = 0
region = europe-west2
[core]
account = aaron.altrock@gmail.com
project = detect-fake-news-313201
[dataproc]
region = europe-west2

Your active configuration is: [default]


## Check the number of files in successive GCP cloud storage buckets

In [96]:
# Count the number of cleaned JSON files from the end of stage 1 in the pipeline
!gsutil ls -l gs://fake_news_cleaned_json/*.json | wc -l

59733


In [97]:
# Count the number of parsed JSON and TTL files into triples at the end of stage 2 in the pipeline
!gsutil ls -l gs://fake_news_ttl_json/*.ttl | wc -l
!gsutil ls -l gs://fake_news_ttl_json/*.json | wc -l

27590
27590


The variance between the 59,733 cleaned files to 27,590 turtle documents would suggest this is due to the raw data containing duplicating records for the same news web page, when the turtles are indexed by the hash value of the URLs and therefore would overwrite leading to small number of samples.

In [98]:
# Based on https://cloud.google.com/bigquery/docs/quickstarts/quickstart-client-libraries#bigquery_simple_app_client-python
from google.cloud import bigquery
client = bigquery.Client(PROJECT)


## Profile the data in its original form held in BigQuery

In [99]:
# BIgQuery data row count
query_job = client.query(
    """
    SELECT COUNT(*) AS POPULATION_COUNT
    FROM `detect-fake-news-313201.fake_news_sql.src_fake_news`
    """
)

res_df = query_job.result().to_dataframe()  # Waits for job to complete.

res_df

POPULATION_COUNT
0             27589

Therefore deviation by one record compared to the number of files in `gs://fake_news_ttl_json`.

In [100]:
# BIgQuery data row count
query_job = client.query(
    """
    WITH URL_LIST AS (
      SELECT 
      URL
      , COUNT(*) AS URL_COUNT
      FROM `detect-fake-news-313201.fake_news_sql.src_fake_news`
      GROUP BY URL
    )
    SELECT * FROM URL_LIST WHERE URL_COUNT > 1
    """
)

res_df = query_job.result().to_dataframe()  # Waits for job to complete.

res_df

Empty DataFrame
Columns: [URL, URL_COUNT]
Index: []

Therefore no samples found to have duplicating URL in the BigQuery table, and all articles have unique URLs.

In [101]:
# BIgQuery data preview
query_job = client.query(
    """
    SELECT *
    FROM `detect-fake-news-313201.fake_news_sql.src_fake_news`
    LIMIT 10
    """
)

res_df = query_job.result().to_dataframe()  # Waits for job to complete.

res_df

url   domain  \
0  http://awm.com/woman-adopts-rescue-dog-starts-...  awm.com   
1  http://awm.com/grand-daughter-sings-favorite-s...  awm.com   
2  http://awm.com/first-class-passenger-sees-a-so...  awm.com   
3  http://awm.com/model-who-boasts-she-transforme...  awm.com   
4  http://awm.com/dad-puts-his-son-into-the-game-...  awm.com   
5  http://awm.com/little-girls-reaction-to-shooti...  awm.com   
6  http://awm.com/out-of-all-the-holiday-recipes-...  awm.com   
7  http://awm.com/a-starbucks-barista-couldnt-fin...  awm.com   
8  http://awm.com/the-easily-offended-whine-when-...  awm.com   
9  http://awm.com/during-heated-love-making-wife-...  awm.com   

                               domain_hash  \
0  domain_9725d0802cf38288710d8bff8f64dcba   
1  domain_9725d0802cf38288710d8bff8f64dcba   
2  domain_9725d0802cf38288710d8bff8f64dcba   
3  domain_9725d0802cf38288710d8bff8f64dcba   
4  domain_9725d0802cf38288710d8bff8f64dcba   
5  domain_9725d0802cf38288710d8bff8f64dcba   
6  domain_9725d0802cf38288710d8bff8f64dcba   
7  domain_9725d0802cf38288710d8bff8f64dcba   
8  domain_9725d0802cf38288710d8bff8f64dcba   
9  domain_9725d0802cf38288710d8bff8f64dcba   

                               title_hash  \
0  title_761e82d05283bff6f1832c7164626e26   
1  title_2a30a9b6a5a75250a7317c60c0f871e6   
2  title_b89b152a0c80235c4e6174a96e98f166   
3  title_9868eac5482fdbfebf7b94494bd078ec   
4  title_08db9d62bccf3a51cb7ded0fdf666568   
5  title_0ef200f6649ab2907625d4db1f796c87   
6  title_0165dfcd088b4df972d8b23ee45198e0   
7  title_55c16d853bfc26cc4697fe2158f2e273   
8  title_ab7f6ea47a85aeacab878345864501df   
9  title_3ed2e9ed1879767b09e706f2a7f5ced5   

                                               title  \
0  Woman Adopts Rescue Dog, Starts Noticing Foul ...   
1  Grand Daughter Sings Favorite Song to Grandma ...   
2  First Class Passenger Sees A Soldier Walking U...   
3  Model Who Boasts She Transformed Into An Afric...   
4  Dad Puts His Son Into The Game Days After Beat...   
5  Little Girl’s Reaction To Shooting Her First D...   
6  Out Of All The Holiday Recipes I’ve Tried, His...   
7  A Starbucks Barista Couldn’t Find A Babysitter...   
8  The Easily Offended Whine When Kelly Clarkson ...   
9  During Heated Love Making, Wife Accidentally B...   

                               body_hash  \
0  body_072ebc1af96077202eb2ad843b6fe4cd   
1  body_3e0c1daeb28fa250067a74240a5af93d   
2  body_6953f3cbe02f2bdf2083937b1c8f9b7f   
3  body_d9e3185ae65d5b97185f08d85c31abec   
4  body_fcb8636125a87f14428856aff72e307f   
5  body_c8d4436c055c1516d67af592b850a17f   
6  body_7a5cb6d058bda9afc409df365438cc7f   
7  body_b2277eca06c8bd84aba7a6512dcbb5e6   
8  body_ddf7af702f3aef83dcb93cc5717ad34f   
9  body_f62f488c90df847302cd010ff91e016b   

                                                body       label  \
0  Kelly Benzel was looking for a new best friend...  unreliable   
1  Dementia is one of the scariest and heartbreak...  unreliable   
2  Air travel can be a tiring and trying experien...  unreliable   
3  Martina Big, a 28-year-old white German model,...  unreliable   
4  A cancer diagnosis never comes at the right ti...  unreliable   
5  In the YouTube video below, you’re going to se...  unreliable   
6  Once a year, the grocery stores start stocking...  unreliable   
7  Starbucks can be a great place to get luxuriou...  unreliable   
8  Dolling out corporal punishment is a controver...  unreliable   
9  He always told her that he liked it rough. But...  unreliable   

                           url_hash  
0  00d76dc2a2b9c02526ab6aba99cea68d  
1  022b655e21ca0d38a1fc33f2a37165a8  
2  0405cdbffb7b0f0aef94eafaba0863ff  
3  09f269f0d217228587029f2012f875c9  
4  0c739d0432f31e7805c0d62f6eff0849  
5  154032234b0955080af9117e817a7676  
6  1af9a9f5326071c7bc5a6e16e04a75df  
7  201402629241e8945bc127ac0b387dc8  
8  21825e99b66d5a6b55e9e3f8c5bdfe71  
9  24c1a494a889e14ef274e14d31fca550

In [102]:
# BIgQuery count by domain
query_job = client.query(
    """
    SELECT
    DOMAIN_HASH
    , LABEL
    , COUNT(*) AS ARTICLES_COUNT
    FROM `detect-fake-news-313201.fake_news_sql.src_fake_news`
    GROUP BY DOMAIN_HASH, LABEL
    ORDER BY ARTICLES_COUNT DESC
    """
)

res_df = query_job.result().to_dataframe()  # Waits for job to complete.

# Tally the domain hash to ensure each domain only has one label
domain_tally_ls = []
duplicate_domain_tally_ls = []
for i, row in res_df.iterrows():
  if row['DOMAIN_HASH'] in domain_tally_ls:
    duplicate_domain_tally_ls += [True]
  else:
    duplicate_domain_tally_ls += [False]
  
  # Add domain hash to the list of domains already reviwed
  domain_tally_ls += [row['DOMAIN_HASH']]

res_df['DOMAIN_HAS_MULTIPLE_LABEL'] = duplicate_domain_tally_ls

res_df

DOMAIN_HASH       LABEL  ARTICLES_COUNT  \
0    domain_8f00b2b61ba335244231d632d390bf8d        fake            7046   
1    domain_9f5bbbcbad4a48edd86162fabbe90b6e   political            5441   
2    domain_7e058b16a00bca2c3620d5147881e34d  conspiracy            2819   
3    domain_98845529bda170844508657ae469197a        bias            1162   
4    domain_2bb13bea21c458421e33179105662bdd   political            1061   
5    domain_bd8db3098e8505e2d80c8b89a2eccdb8   clickbait             813   
6    domain_0d6ca907eef322628f81ae5a38735d44        bias             730   
7    domain_f573aed8d0dcf08ef6ce3efd5f5c48c0   political             729   
8    domain_213f2e7f6968dd534eb43ef113fce3e1   political             671   
9    domain_8fbb80fdc3a337fd46babecb01f203fa     junksci             615   
10   domain_799497bdca60cb2f1522d66166f15274   political             435   
11   domain_bfcde262dcc0a3fe874d75293764fb4c   political             415   
12   domain_69a49c3be87af6ae0c48f5c9b005a21b        None             413   
13   domain_f9f6f6f07772425d1522dd8a8ab25193        None             347   
14   domain_5dfa617b9d6dac8de770c521298c4e53     junksci             294   
15   domain_3cbda5292ac5c0b572a8d764b0e52eec        bias             209   
16   domain_fb4e14ee232e6a6acfda1de4f10a1fe3   political             209   
17   domain_c29df259d72ab43dc7e2c43b1941a1c9   clickbait             178   
18   domain_6a12243e8071c910a53039000309c901    reliable             168   
19   domain_7cc6378eb1ac3bf4455eb536225df80b        bias             160   
20   domain_f9e7e3cf456dda4da1f7ada4ba6ddc8c  conspiracy             153   
21   domain_191e287683c85a13fa54c85c43281769  conspiracy             152   
22   domain_3e2b4bc8d67cf01a6f156413acc03375   clickbait             144   
23   domain_1682643b797f67fe89d09642a50174fe   political             120   
24   domain_b55734eb636ef77462a42b666bc74c5f     unknown             114   
25   domain_714bf72800dbdc780c8b895adc12ae3c  unreliable             108   
26   domain_bfff154f86499943be9f3e65798b5d53   political             107   
27   domain_d061e62c09e8c363752ddfa81c54b3ba        bias              99   
28   domain_a66294ec7d2d0d8e3b22c293a88d14f4   political              97   
29   domain_8776e5a68ce1577dd2809b859c3c6d82   political              90   
30   domain_5014c2313218bc048eaef8f34771d2eb     unknown              84   
31   domain_f03bcedc7a83b5568ff986bfec2c3413  conspiracy              77   
32   domain_9725d0802cf38288710d8bff8f64dcba  unreliable              76   
33   domain_fa702aa239d1e1c37cc9c5d75ab9b4f9   political              75   
34   domain_f22938424edda66ec091ddd069854179  unreliable              74   
35   domain_77b6aa19bdd7ac5f26e593472bd9a7d5        hate              71   
36   domain_6afcd6f29c592f5495589242b12d1c25   political              67   
37   domain_2c6ca1a760cdc4987f4283c9464dc639        None              63   
38   domain_090b7aeff850c5dda848d5b9e75bb16d     unknown              58   
39   domain_f047cbddff341a471b7b05bd5f4ee660       rumor              57   
40   domain_daf12fe2bd3c219af940a4f8891f23f0        bias              57   
41   domain_bc6e2fd29bb66b366afa86a3934609ce  conspiracy              54   
42   domain_dbbb88dd5835637c6c4b2e3c54c2669e   political              50   
43   domain_b04f082dca5817be6d561c1e25ea2c16        None              48   
44   domain_b2f824ff829b40551a0f7e99a53745a0      satire              47   
45   domain_e23a8807f4586e5ab98d2c0d335d998a  conspiracy              45   
46   domain_9c34eb473390528f31c0303bda283c21  unreliable              44   
47   domain_95e62162e78b589ec6362d5a6a19b3cb        bias              43   
48   domain_b0c5e6bffbaa50cb2a380e69d80f0fde        bias              42   
49   domain_c072fb2a857d0288f93a765a423629c1  conspiracy              42   
50   domain_9a0d9e918e815c1424ffff912643042f  unreliable              41   
51   domain_f3408542d52cb5a4c7f9d764ba80713b        bias              41   
52   

In [79]:
# BIgQuery count by label
query_job = client.query(
    """
    SELECT
    LABEL
    , COUNT(*) AS LABEL_COUNT
    FROM `detect-fake-news-313201.fake_news_sql.src_fake_news`
    GROUP BY LABEL
    ORDER BY LABEL_COUNT DESC
    """
)

res_df = query_job.result().to_dataframe()

print('Total: {}'.format(res_df['LABEL_COUNT'].sum()))

res_df

Total: 27589


LABEL  LABEL_COUNT
0    political         9776
1         fake         7081
2   conspiracy         3512
3         bias         2793
4    clickbait         1238
5      junksci          993
6         None          990
7   unreliable          437
8      unknown          354
9     reliable          178
10      satire           99
11        hate           81
12       rumor           57

In [106]:
# BIgQuery list of URLs
query_job = client.query(
    """
    SELECT DISTINCT
    URL_HASH
    FROM `detect-fake-news-313201.fake_news_sql.src_fake_news`
    """
)

res_df = query_job.result().to_dataframe()

print('Total: {}'.format(res_df.shape[0]))

res_df.head()

Total: 27589


URL_HASH
0  00d76dc2a2b9c02526ab6aba99cea68d
1  022b655e21ca0d38a1fc33f2a37165a8
2  0405cdbffb7b0f0aef94eafaba0863ff
3  09f269f0d217228587029f2012f875c9
4  0c739d0432f31e7805c0d62f6eff0849

Noted that there were no classification for 990 samples, and further 354 with unknown classifications.



## Profile the data in GraphDB

In [80]:
# Install the wrapper package
# Source: https://github.com/RDFLib/sparqlwrapper
!pip install sparqlwrapper

In [110]:
# Code based on: https://sparqlwrapper.readthedocs.io/en/latest/main.html
from SPARQLWrapper import SPARQLWrapper, JSON

# queryString = """
# PREFIX aa: <http://www.city.ac.uk/ds/inm363/aaron_altrock#>
# PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
# PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

# select ?a ?b ?c where {
#   ?a ?b ?c .
#   ?a rdf:type aa:urlHash .
#   aa:unreliable rdf:type aa:newsLabel .
#   aa:unreliable rdfs:label "unreliable" .
# } limit 100
# """

queryString = """
PREFIX aa: <http://www.city.ac.uk/ds/inm363/aaron_altrock#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

select (count(?url_hash) as ?url_count) where {
  ?url_hash rdf:type aa:urlHash .
}
"""


sparql = SPARQLWrapper("http://35.246.120.165:7200/repositories/src_fake_news")
sparql.setReturnFormat(JSON)
sparql.setQuery(queryString)

try :
   res_dct = sparql.query().convert()
   print('OK')

except Exception as e:
   print('ERROR: {}'.format(e))
   

OK


In [117]:
# No. of URL hash in GraphDB
res_dct.get('results').get('bindings')[0].get('url_count').get('value')

'25211'

In [83]:
pd.DataFrame.from_dict(res_dct.get('results').get('bindings'), orient='columns')

a  \
0   {'type': 'uri', 'value': 'http://www.city.ac.u...   
1   {'type': 'uri', 'value': 'http://www.city.ac.u...   
2   {'type': 'uri', 'value': 'http://www.city.ac.u...   
3   {'type': 'uri', 'value': 'http://www.city.ac.u...   
4   {'type': 'uri', 'value': 'http://www.city.ac.u...   
5   {'type': 'uri', 'value': 'http://www.city.ac.u...   
6   {'type': 'uri', 'value': 'http://www.city.ac.u...   
7   {'type': 'uri', 'value': 'http://www.city.ac.u...   
8   {'type': 'uri', 'value': 'http://www.city.ac.u...   
9   {'type': 'uri', 'value': 'http://www.city.ac.u...   
10  {'type': 'uri', 'value': 'http://www.city.ac.u...   
11  {'type': 'uri', 'value': 'http://www.city.ac.u...   
12  {'type': 'uri', 'value': 'http://www.city.ac.u...   
13  {'type': 'uri', 'value': 'http://www.city.ac.u...   
14  {'type': 'uri', 'value': 'http://www.city.ac.u...   
15  {'type': 'uri', 'value': 'http://www.city.ac.u...   
16  {'type': 'uri', 'value': 'http://www.city.ac.u...   
17  {'type': 'uri', 'value': 'http://www.city.ac.u...   
18  {'type': 'uri', 'value': 'http://www.city.ac.u...   
19  {'type': 'uri', 'value': 'http://www.city.ac.u...   
20  {'type': 'uri', 'value': 'http://www.city.ac.u...   
21  {'type': 'uri', 'value': 'http://www.city.ac.u...   
22  {'type': 'uri', 'value': 'http://www.city.ac.u...   
23  {'type': 'uri', 'value': 'http://www.city.ac.u...   
24  {'type': 'uri', 'value': 'http://www.city.ac.u...   
25  {'type': 'uri', 'value': 'http://www.city.ac.u...   
26  {'type': 'uri', 'value': 'http://www.city.ac.u...   
27  {'type': 'uri', 'value': 'http://www.city.ac.u...   
28  {'type': 'uri', 'value': 'http://www.city.ac.u...   
29  {'type': 'uri', 'value': 'http://www.city.ac.u...   
30  {'type': 'uri', 'value': 'http://www.city.ac.u...   
31  {'type': 'uri', 'value': 'http://www.city.ac.u...   
32  {'type': 'uri', 'value': 'http://www.city.ac.u...   
33  {'type': 'uri', 'value': 'http://www.city.ac.u...   
34  {'type': 'uri', 'value': 'http://www.city.ac.u...   
35  {'type': 'uri', 'value': 'http://www.city.ac.u...   
36  {'type': 'uri', 'value': 'http://www.city.ac.u...   
37  {'type': 'uri', 'value': 'http://www.city.ac.u...   
38  {'type': 'uri', 'value': 'http://www.city.ac.u...   
39  {'type': 'uri', 'value': 'http://www.city.ac.u...   
40  {'type': 'uri', 'value': 'http://www.city.ac.u...   
41  {'type': 'uri', 'value': 'http://www.city.ac.u...   
42  {'type': 'uri', 'value': 'http://www.city.ac.u...   
43  {'type': 'uri', 'value': 'http://www.city.ac.u...   
44  {'type': 'uri', 'value': 'http://www.city.ac.u...   
45  {'type': 'uri', 'value': 'http://www.city.ac.u...   
46  {'type': 'uri', 'value': 'http://www.city.ac.u...   
47  {'type': 'uri', 'value': 'http://www.city.ac.u...   
48  {'type': 'uri', 'value': 'http://www.city.ac.u...   
49  {'type': 'uri', 'value': 'http://www.city.ac.u...   
50  {'type': 'uri', 'value': 'http://www.city.ac.u...   
51  {'type': 'uri', 'value': 'http://www.city.ac.u...   
52  {'type': 'uri', 'value': 'http://www.city.ac.u...   
53  {'type': 'uri', 'value': 'http://www.city.ac.u...   
54  {'type': 'uri', 'value': 'http://www.city.ac.u...   
55  {'type': 'uri', 'value': 'http://www.city.ac.u...   
56  {'type': 'uri', 'value': 'http://www.city.ac.u...   
57  {'type': 'uri', 'value': 'http://www.city.ac.u...   
58  {'type': 'uri', 'value': 'http://www.city.ac.u...   
59  {'type': 'uri', 'value': 'http://www.city.ac.u...   
60  {'type': 'uri', 'value': 'http://www.city.ac.u...   
61  {'type': 'uri', 'value': 'http://www.city.ac.u...   
62  {'type': 'uri', 'value': 'http://www.city.ac.u...   
63  {'type': 'uri', 'value': 'http://www.city.ac.u...   
64  {'type': 'uri', 'value': 'http://www.city.ac.u...   
65  {'type': 'uri', 'value': 'http://www.city.ac.u...   
66  {'type': 'uri', 'value': 'http://www.city.ac.u...   
67  {'type': 'uri', 'value': 'http://www.city.ac.u...   
68  {'type': 'uri', 'value': 'http://www.city.ac.u...   
69  {'type': 'uri', 'value': 'http://www.city.ac.u...   
70  {